src/algorithms/convex.rs

```diff
@@ -0,0 +1,229 @@
+//! Convexity checking for portgraphs.
+//!
+//! This is based on a [`ConvexChecker`] object that is expensive to create
+//! (linear in the size of the graph), but can be reused to check multiple
+//! subgraphs for convexity quickly.
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use crate::algorithms::toposort;
+use crate::{Direction, LinkView, NodeIndex, PortIndex};
+
+use super::TopoSort;
+
+#[derive(Default, Clone, Debug, PartialEq, Eq)]
+enum Causal {
+    #[default]
+    P, // in the past
+    F, // in the future
+}
+
+/// A pre-computed datastructure for fast convexity checking.
+///
+/// TODO: implement for graph traits?
+pub struct ConvexChecker<G> {
+    graph: G,
+    // The nodes in topological order
+    topsort_nodes: Vec<NodeIndex>,
+    // The index of a node in the topological order (the inverse of topsort_nodes)
+    topsort_ind: BTreeMap<NodeIndex, usize>,
```

Comment:
The indices should be dense on a full graph toposort, so an `UnmanagedMap` would do better here.

Reply:
Done